### PR TITLE
[SPARK-23911][SQL] Add aggregate function.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

This pr adds `aggregate` function which applies a binary operator to an initial state and all elements in the array, and reduces this to a single state. The final state is converted into the final result by applying a finish function.

```sql
> SELECT aggregate(array(1, 2, 3), (acc, x) -> acc + x);
 6
> SELECT aggregate(array(1, 2, 3), (acc, x) -> acc + x, acc -> acc * 10);
 60
```

## How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @hvanhovell @gatorsmile @cloud-fan

**Comment by mn-mikke** [comment]
Isn't this PR related to the Jira ticket [SPARK-23911](https://issues.apache.org/jira/browse/SPARK-23911)?

**Comment by ueshin** [comment]
Oops, yes, I wrote a wrong jira-id. Fixed. Thanks!

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
@ueshin  You need to address the conflicts again. :)

**Comment by ueshin** [comment]
Thanks! merging to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala [hunk 1]
```diff
@@ -270,3 +278,90 @@ case class ArrayFilter(
...
+@ExpressionDescription(
+  usage =
+    """
+      _FUNC_(expr, start, merge, finish) - Applies a binary operator to an initial state and all
+      elements in the array, and reduces this to a single state. The final state is converted
+      into the final result by applying a finish function.
+    """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), (acc, x) -> acc + x);
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala]: @ueshin, would you mind if I ask to kindly double check if the example works? seems not in my local:

```
spark-sql> SELECT aggregate(array(1, 2, 3), (acc, x) -> acc + x);
2018-08-08 16:08:25 ERROR SparkSQLDriver:91 - Failed in [SELECT aggregate(array(1, 2, 3), (acc, x) -> acc + x)]
org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to dataType on unresolved object, tree: 'acc
	at org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute.dataType(unresolved.scala:105)
	at org.apache.spark.sql.catalyst.expressions.BinaryArithmetic.dataType(arithmetic.scala:119)
	at org.apache.spark.sql.catalyst.expressions.LambdaFunction.dataType(higherOrderFunctions.scala:72)
	at org.apache.spark.sql.hive.HiveSessionCatalog$$anonfun$1.apply(HiveSessionCatalog.scala:122)
	at org.apache.spark.sql.hive.HiveSessionCatalog$$anonfun$1.apply(HiveSessionCatalog.scala:121)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
```

Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala]: Oops, sorry, we need the second argument as an initial value.

```sql
SELECT aggregate(array(1, 2, 3), 0, (acc, x) -> acc + x);
```

I'll submit a follow-up pr soon.

**Pull request closed without merging**